test/es-module/test-esm-dynamic-import.js

```diff
@@ -0,0 +1,66 @@
+// Flags: --experimental-modules --harmony-dynamic-import
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const { URL } = require('url');
+const vm = require('vm');
+
+common.crashOnUnhandledRejection();
+
+const relativePath = './test-esm-ok.mjs';
+const absolutePath = require.resolve('./test-esm-ok.mjs');
+const targetURL = new URL('file:///');
+targetURL.pathname = absolutePath;
+
+import(relativePath)
+  .then(common.mustCall(ns => {
+    // Can't deepStrictEqual because ns isn't a normal object
+    assert.deepEqual(ns, { default: true });
+  }));
+
+import("fs")
+  .then(common.mustCall(ns => {
+    assert.equal(typeof ns.default.writeFile, 'function');
+  }));
+
+// Succeeds
+Promise.resolve(vm.runInThisContext('import("fs")'))
+  .then(common.mustCall(ns => {
+    assert.equal(typeof ns.default.writeFile, 'function');
+  }));
+
+// Succeeds because it's got an valid base url
+Promise.resolve(vm.runInThisContext(`import("${relativePath}")`, {
```

Comment:
Note: This will introduce a de-facto global access to `require` that doesn't exist right now. E.g. random code running in `vm.runInThisContext` will gain access to all node APIs.

Before:

```bash
% ./node <(echo "console.log(require('vm').runInThisContext('require(\"fs\")'));")
evalmachine.<anonymous>:1
require("fs")
^

ReferenceError: require is not defined
# ...
```

After:

```bash
% ./node --experimental-modules --harmony <(echo "require('vm').runInThisContext('import(\"fs\")').then(console.log);")
(node:29002) ExperimentalWarning: The ESM module loader is experimental.
{ default:
   { constants:
# ...
```